src/gfn/examples/distributions.py

```diff
@@ -0,0 +1,221 @@
+        assert alpha.shape == (self.n_states, self.n_components)
+        assert beta.shape == (self.n_states, self.n_components)
+
+        self.base_dist = MixtureSameFamily(
+            Categorical(logits=mixture_logits),
+            Beta(alpha, beta),
+        )
+
+        self.min_angles, self.max_angles = self.get_min_and_max_angles()
+
+    def get_min_and_max_angles(self):
+        if self.northeastern:
+            min_angles = torch.where(
+                self.centers[:, 0] <= 1 - self.delta,
+                0.0,
+                2.0 / torch.pi * torch.arccos((1 - self.centers[:, 0]) / self.delta),
+            )
+            max_angles = torch.where(
+                self.centers[:, 1] <= 1 - self.delta,
+                1.0,
+                2.0 / torch.pi * torch.arcsin((1 - self.centers[:, 1]) / self.delta),
+            )
+        else:
+            min_angles = torch.where(
+                self.centers[:, 0] >= self.delta,
+                0.0,
+                2.0 / torch.pi * torch.arccos((self.centers[:, 0]) / self.delta),
+            )
+            max_angles = torch.where(
+                self.centers[:, 1] >= self.delta,
+                1.0,
+                2.0 / torch.pi * torch.arcsin((self.centers[:, 1]) / self.delta),
+            )
+
+        return min_angles, max_angles
+
+    def rsample(self, sample_shape=()):
+        base_01_samples = self.base_dist.rsample(sample_shape=sample_shape)
+
+        sampled_angles = (
+            self.min_angles + (self.max_angles - self.min_angles) * base_01_samples
+        )
+        sampled_angles = torch.pi / 2 * sampled_angles
+
+        sampled_actions = (
+            torch.stack(
+                [torch.cos(sampled_angles), torch.sin(sampled_angles)],
+                dim=1,
+            )
+            * self.delta
+        )
+
+        if self.northeastern:
+            sampled_next_states = self.centers + sampled_actions
+        else:
+            sampled_next_states = self.centers - sampled_actions
+        # TODO: do we need to return actions or next states here? -- logprobs are ok, but the previous 4 lines might not be
+        return sampled_next_states
+
+    def log_prob(self, sampled_next_states):
+        if self.northeastern:
+            sampled_actions = sampled_next_states - self.centers
+        else:
+            sampled_actions = self.centers - sampled_next_states
+
+        sampled_angles = torch.arccos(sampled_actions[:, 0] / self.delta)
+
+        sampled_angles = sampled_angles / (torch.pi / 2)
+
+        base_01_samples = (sampled_angles - self.min_angles) / (
+            self.max_angles - self.min_angles
+        )
+
+        base_01_logprobs = self.base_dist.log_prob(base_01_samples)
+
+        logprobs = (
+            base_01_logprobs
+            - np.log(self.delta)
+            - np.log(np.pi / 2)
+            - torch.log(self.max_angles - self.min_angles)
+        )
+
+        return logprobs
```

Comment:
I find these two implementations extremely clean. For this example we could comment liberally or make explicit the math in the form of equations - but I like the work flow of `next_states = obj.sample()` -> `logprobs = obj.calc_logprob(next_states)`.

Reply:
Right, there is some math in the paper justifying this code. I'll incorporate some of it in the comments.